### ISSUE TITLE
Implement a `fill(p)` prequel for every solver, where `p` might be a number.

### ISSUE THREAD

**Issue by kellertuer** [JuliaManifolds/Manopt.jl#235]
Currently we have that the last line in 

```julia
using Manifolds
M = Circle()
p = 0.
X = π/2
q = exp(M,p,X)
exp!(M, q, p, X)
```

fails since that function does not exist for Circle (or `Euclidean()` or other number-type manifolds, I have a few listed at
https://github.com/JuliaManifolds/Manopt.jl/blob/9fc85ccf89f3aacbddcb57e3144fbe69fa23f62f/src/Manopt.jl#L193
This makes generic algorithms for them complicated, indeed quite a while ago I started 
https://github.com/JuliaManifolds/Manopt.jl/blob/master/src/plans/nonmutating_manifolds_plans.jl
which is only loaded if you load manifolds together with Manopt.jl

But that is code duplication (besides using `retract` in all places instead of `retract!`).

What do you think about introducing `retract!`? Is that possible or would that break stuff? We can not modify `q` sure, but can we find some other trick?

**Comment by kellertuer** [comment]
Since it is similar (working through all solvers) this could be done together with #223.

**Comment by kellertuer** [comment]
There is one caveat here (and a way to avoid one copy)
the actual high level interface might return more than just the result `q` (when setting `return_state=true`. Besides that with the filled value one can also directly call `gradient_descent!`. So this would look more like

```julia
function gradient_descent(M::AbstractManifold, f, grad_f, p::Number, kwargs...)
    q = fill(p)
    rs = gradient_descent!(M, f, grad_f, q, kwargs...)
    #return just a number if  the return type is the same as the type of q
    (typeof(q) == typeof(rs)) && (return rs[])
    return rs 
end
```

this is still a quite generic way we just have to intrdoduce to every solver (like currently the `gradient_descent` non-mutating ones are very boiler-plate-ly).

**Comment by kellertuer** [comment]
Hm, it is a bit more complicated, since `grad_f` _also_ returns numbers and has to be adopted. I will slowly continue on this but it might actually take quite some time.

**Comment by mateuszbaran** [comment]
Hm, right, that's actually more complicated than I thought initially.

**Comment by kellertuer** [comment]
Well, I got it to work and now know what to avoid.
The current state is far less code than the old reimplementation of Linesearch; so that's nice.

We should just not use `fill` on Floats it seems.

**Comment by kellertuer** [comment]
This will be resolved in #236.
